File: examples/python/04_streaming/main.py
"""Stream stream.delta tokens to stdout in real time."""

from __future__ import annotations

import asyncio
import os
import sys
from pathlib import Path

from ai_agent import Agent, AgentConfig, StreamingConfig

REPO_ROOT = Path(__file__).resolve().parents[3]
AGENT_BINARY = str(REPO_ROOT / "agent")

ENV = {
    "SLLM_ENDPOINT": os.getenv("SLLM_ENDPOINT", "http://localhost:8080/v1/chat/completions"),
    "SLLM_API_KEY": os.getenv("SLLM_API_KEY", "sk-gemma4"),
}


def on_delta(text: str, turn: int) -> None:
    # Print deltas inline — turn number is useful for multi-turn flows.
    sys.stdout.write(text)
    sys.stdout.flush()


def on_status(ratio: float, count: int, limit: int) -> None:
    # Render context usage to stderr so it doesn't mix with the streamed answer.
    print(f"\n[ctx {count}/{limit} = {ratio:.0%}]", file=sys.stderr)


async def main() -> None:
    async with Agent(binary_path=AGENT_BINARY, env=ENV) as agent:
        # Streaming must be enabled before deltas are emitted.
        await agent.configure(AgentConfig(
            max_turns=4,
            streaming=StreamingConfig(enabled=True, context_status=True),
        ))
        result = await agent.run(
            "100文字くらいで自己紹介して。",
            stream=on_delta,
            on_status=on_status,
        )
        print("\n---")
        print("final reason:", result.reason, "| turns:", result.turns)


if __name__ == "__main__":
    asyncio.run(main())

File: examples/js/04_streaming/main.ts
/** Use runStream() AsyncIterable to render deltas as they arrive. */

import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';

import { Agent } from '@ai-agent/sdk';

const HERE = dirname(fileURLToPath(import.meta.url));
const AGENT_BINARY = resolve(HERE, '../../../agent');

const env = {
  SLLM_ENDPOINT: process.env.SLLM_ENDPOINT ?? 'http://localhost:8080/v1/chat/completions',
  SLLM_API_KEY: process.env.SLLM_API_KEY ?? 'sk-gemma4',
};

async function main(): Promise<void> {
  await using agent = await Agent.open({ binaryPath: AGENT_BINARY, env });
  await agent.configure({
    max_turns: 4,
    streaming: { enabled: true, context_status: true },
  });

  // for-await over StreamEvent: 'delta' | 'status' | 'end'.
  for await (const ev of agent.runStream('100文字くらいで自己紹介して。')) {
    if (ev.kind === 'delta') {
      process.stdout.write(ev.text);
    } else if (ev.kind === 'status') {
      process.stderr.write(
        `\n[ctx ${ev.tokenCount}/${ev.tokenLimit} = ${(ev.usageRatio * 100).toFixed(0)}%]\n`,
      );
    } else if (ev.kind === 'end') {
      console.log('\n---');
      console.log('final reason:', ev.result.reason, '| turns:', ev.result.turns);
    }
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
